src/core/dense_set.cc

```diff
@@ -0,0 +1,558 @@
+  la.construct(lk);
+
+  lk->next = *it;
+  it->SetLink(lk);
+  lk->SetObject(data);
+
+  return ObjectAllocSize(data);
+}
+
+void DenseSet::PushFront(DenseSet::ChainVectorIterator it, DenseSet::DensePtr* ptr) {
+  if (it->IsEmpty()) {
+    it->SetObject(ptr->GetObject());
+    if (ptr->IsLink()) {
+      // deallocate the link if it is no longer a link as it is now in an empty list
+      mr()->deallocate(ptr->Raw(), sizeof(DenseLinkKey), alignof(DenseLinkKey));
+    }
+  } else if (ptr->IsLink()) {
+    // if the pointer is already a link then no allocation needed
+    *ptr->Next() = *it;
+    *it = *ptr;
+  } else {
+    // allocate a new link if needed and copy the pointer to the new link
+    LinkAllocator la(mr());
+    DenseLinkKey* lk = la.allocate(1);
+    la.construct(lk);
+
+    lk->next = *it;
+    lk->SetObject(ptr->GetObject());
+    it->SetLink(lk);
+  }
+}
+
+void* DenseSet::PopFront(DenseSet::ChainVectorIterator it) {
+  if (IsEmpty(it)) {
+    return nullptr;
+  }
+
+  void* ret = it->GetObject();
+
+  // if this is the last object in this chain just reset the
+  // DensePtr instead of deleting it
+  if (it->IsObject()) {
+    it->Reset();
+  } else {
+    DCHECK(it->IsLink());
+    // set the current item in the list to the next one & deallocate the link data structure
+    DenseLinkKey* lk = (DenseLinkKey*)it->Raw();
+    *it = lk->next;
+    mr()->deallocate(lk, sizeof(DenseLinkKey), alignof(DenseLinkKey));
+  }
+
+  return ret;
+}
+
+DenseSet::DensePtr* DenseSet::Find(void* needle, DenseSet::ChainVectorIterator list,
+                                   DenseSet::DensePtr** prev_out) {
+  DensePtr* prev = nullptr;
+  ChainIterator it = ChainBegin(list);
+  for (; it != ChainEnd(list); ++it) {
+    if (Equal(*it, needle)) {
+      break;
+    }
+
+    prev = *it;
+  }
+
+  if (!it->IsEmpty() && prev_out != nullptr) {
+    *prev_out = prev;
+  }
+
+  return *it;
+}
+
+DenseSet::ChainIterator DenseSet::Unlink(DenseSet::ChainIterator node,
+                                          DenseSet::ChainIterator prev, DenseSet::DensePtr* out) {
+  DCHECK(out != nullptr);
+  if (node->IsObject()) {
+    // this is the only item in the list, reset it and keep the iterator here
+    if (prev == nullptr) {
+      // copy the pointer to the data before a reset
+      *out = **node;
+      node->Reset();
+      return node;
+    } else {
+      DCHECK(prev->IsLink());
+      /*
+        here we can try to pull a sneaky optimization, since the caller
+        may get a LinkKey or an regular object it is responsible for freeing it.
+        Some use cases (i.e Grow) may be able to use a LinkKey without freeing it as they can be
+        inserted into the front of a non-empty chained list. So instead of freeing them here
+        and save some memory management calls
+      */
+      DenseLinkKey* lk = (DenseLinkKey*)prev->Raw();
+      DCHECK(!lk->IsDisplaced());
+
+      // swap the two pointers to the actual data
```

Comment:
I do not understand this double swap.
why not just copy from `**node` to `*out` and then link prev to `node->next`?

Reply:
If I'm understanding correctly the issue I ran into with that approach was that in this control branch we are unlinking the end of a chain so effectively `node` points to the `DensePtr next` in the previous `DenseLinkKey` so `node->next` would be a `nullptr` and I wanted to avoid having a `DenseLinkKey` at the end of a chain with a `nullptr` as next to save space as well as the optimization of not having to deallocate / reallocate the link on a grow in this case.

Looking back now this was definitely a premature optimization which was not worth the drop in code quality.